src/blr/cli/tagfastq.py

```diff
@@ -0,0 +1,132 @@
+
+logger = logging.getLogger(__name__)
+
+
+def main(args):
+    logger.info(f'Starting')
+
+    # Get the corrected barcodes and create a dictionary pointing each raw barcode to its canonical sequence.
+
+    with open(args.barcodes_corrected, "r") as reader:
+        corrected_barcodes = parse_corrected_barcodes(reader)
+
+    # Get the raw barcodes and create a dictionary pointing each read header to the raw barcode sequence.
+    with dnaio.open(args.barcodes_raw, mode="r") as reader:
+        header_barcodes = parser_raw_barcodes(reader)
+
+    in_interleaved = True if not args.input2 else False
+    logger.info(f"Input detected as {'interleaved fastq.' if in_interleaved else 'paired fastq.'}")
+
+    # If no output1 is given output is sent to stdout
+    if not args.output1:
+        logger.info(f"Writing output to stdout.")
+        args.output1 = sys.stdout.buffer
+        args.output2 = None
+
+    out_interleaved = True if not args.output2 else False
+    logger.info(f"Output detected as {'interleaved fastq.' if out_interleaved else 'paired fastq.'}")
+
+    reads_skipped = list()
+    # Parse input FASTA/FASTQ for read1 and read2 and write output
+    with dnaio.open(args.input1, file2=args.input2, interleaved=in_interleaved, mode="r") as reader, \
+            dnaio.open(args.output1, file2=args.output2, interleaved=out_interleaved, mode="w") as writer:
+        for read1, read2 in tqdm(reader, desc='Read pairs processed'):
+            # Header parsing
+            name_and_pos_r1, read_and_index_r1 = read1.name.split(maxsplit=1)
+            name_and_pos_r2, read_and_index_r2 = read2.name.split(maxsplit=1)
+
+            assert name_and_pos_r1 == name_and_pos_r2
+
+            try:
+                raw_barcode_seq = header_barcodes[name_and_pos_r1]
+            except KeyError:
+                reads_skipped.append(name_and_pos_r1)
+                continue
+
+            corr_barcode = corrected_barcodes[raw_barcode_seq]
+
+            # Make new string with barcode sequence and id to add to headers.
+            new_text = f"{corr_barcode.seq}_{args.barcode_cluster_tag}:Z:{corr_barcode.id}"
+
+            # Save header to read instances
+            read1.name = f"{name_and_pos_r1}_{new_text} {read_and_index_r1}"
+            read2.name = f"{name_and_pos_r2}_{new_text} {read_and_index_r2}"
+
+            # Write to out
+            writer.write(read1, read2)
+
+    for header in reads_skipped:
+        logger.info(f"Barcode for read-pair {header} not found. Skipped read-pair.")
+
+    logger.info(f'Finished')
+
+
+def parse_corrected_barcodes(open_file):
+    """
+    Parse starcode cluster output and return a dictionary with raw sequences pointing to a
+    corrected canonical sequence
+    :param open_file: starcode tabular output file.
+    :return: dict: raw sequences pointing to a nametuple containe corrected canonical sequence and id.
+    """
+    target = namedtuple("Cluster", ['seq', 'id'])
```

Comment:
> It took me a little while to figure this out - since we're dealing with both reads names and cluster integers here, `id` was a little big ambiguous. Maybe we could substitute it with `fastq_header` or something? I am mainly thinking about it being easier to understand in the future.

I can maybe change it to `index`for now as I think this better describes it. The only reason this is included is that we have been using these cluster integer numbers to represent the "canonical sequence" that is given by starcode, idealy I think we should move away from this integer entirely and just use the canonical sequence directly. This might however interfere with some thing we have written earlier so we have to check how to do the change. Also, to use the barcode sequence is standard for most 10x or other linked-read related tools I have seen so this would standardize our process a bit more